### ISSUE TITLE
问题

### ISSUE THREAD

**Issue by jujujujuz** [mcpaulgeorge/WalMaFa#3]
你好作者，有一些问题希望你能帮我解答
<img width="584" alt="30a4b3d6cec5deb2c8961f6fb29ddff" src="https://github.com/user-attachments/assets/e8b833e1-ff35-472c-9628-2eaa31fe414b">
在你的论文中，损失函数是这样定义的。
而在你的代码中损失函数是这样的
`        content_loss = L1loss(restored, target)
        ssim_loss = 1 - ssim(restored, target, data_range=1.0).to("cuda:0")
        content_loss = content_loss+ssim_loss
        tv_loss = TVLoss()
        TV_loss = tv_loss(restored)
        loss = 0.95*content_loss + 0.05*TV_loss`

我并没有找到在你论文中提到的，有关小波变换和傅里叶变换的损失函数。是并没有用到，还是在哪里我没有找到，希望你能帮我解答。

还有一个问题就是关于WMB模块的问题
<img width="316" alt="b7d61caab5d972500513ae2cb44091a" src="https://github.com/user-attachments/assets/135c93e4-6261-4e32-8df2-eaac581c14b4">
在你的论文模块是这样的
而在代码里面模块是这样
`class WMB(nn.Module):
    def __init__(self, dim, num_heads=1, ffn_expansion_factor=2.66, bias=True, LayerNorm_type='WithBias'):
        super(WMB, self).__init__()
        self.DWT = blocks.DWT()
        self.IWT = blocks.IWT()
        self.norm1 = LayerNorm(dim, LayerNorm_type)
        self.illu = Illumination_Estimator(dim, n_fea_in=dim+1, n_fea_out=dim)
        self.norm2 = LayerNorm(dim, LayerNorm_type)
        self.ffn = FeedForward(dim, ffn_expansion_factor, bias)
        self.mb = WM(dim)

    def forward(self, input_):
        global m
        x = input_
        n, c, h, w = x.shape
        x = self.norm1(x)
        x = data_transform(x)
        input_dwt = self.DWT(x)
        # input_LL=A [B,C,H/2,W/2]   input_high0={V,H,D} [3B,C,H/2,W/2]
        input_LL, input_high = input_dwt[:n, ...], input_dwt[n:, ...]
        input_LL, input_image = self.illu(input_LL)
        input_high = self.mb(input_high)

        output = self.IWT(torch.cat((input_LL, input_high), dim=0))
        output = inverse_data_transform(output)

        x = x + output
        x = x + self.ffn(self.norm2(x))
        return x`

按在论文中图的解释，低频信息要经过mamba，高频信息只用过卷积就可以了。而在代码中则反了过来。希望你能帮我解释。

**Comment by mcpaulgeorge** [comment]
github的代码相对于论文内容有些改动。以github代码为主。

Loss函数我们还是选用的L1和SSIM的结合加上TVLoss平衡预测后的过曝噪声，常规的小波和傅里叶的loss会出现振荡，所以没有选取，而TVloss是对原预测图而言的，不会因gt影响。     
Mamba位置问题：WalMaFa模型中的WMB中运用Channel-Mamba主要是起到补充通道空间信息的作用，经过我们进一步实验得知，通道空间特征对高频的贡献大于低频。其主要原因是，高频信息由原始特征图上的三个采样方向(垂直，水平，对角线)组成，主要由边缘等信息为主，而低频主要代表图像整体的颜色等信息，已经蕴含了大部分的通道信息，因此高频上更缺乏通道间信息的整合。当然，您也可以尝试在低频和高频上都添加channel-mamba，看看性能是否有改善。
